### DISCUSSION TITLE
How to configure Trino to use ofs

### DISCUSSION THREAD

**Discussion by arghya18** [apache/ozone#5004] [FAQ]
Anyone configured Trino to use ofs ?
Currently we get below error

Caused by: org.apache.hadoop.fs.UnsupportedFileSystemException: No FileSystem for scheme "ofs"

**Comment by cloudcheflabs** [comment]
I got the same issue. I tried to add ozone fs jar to trino plugin directory, for instance, iceberg, but it failed and it seems that trino does not support ozone ofs.

So, I have tried another alternative. 
Ozone provides s3g gateway. 
After I made ozone secure with kerberos authentication, I can get access and secret key to access ozone via s3g.
I added for instance, iceberg catalog with s3 credential properties to trino as usual, and then, I can access ozone with trino using s3g now.
I hope, it helps you.

- Kidong.

**Reply by arghya18** [reply]
Yes, s3g works but is not performant when compared to ofs. Its not working for us in production.

**Reply by cloudcheflabs** [reply]
@arghya18 
Thank you for your reply. 
I am wondering your setups for s3g performant test, because I have not tested the performance between s3g and ofs for now.
You have compared the performance between s3g and ofs. I want to know your setup to measure the performance, for example.
- how many s3gs are running in your ozone cluster?
- have you used to proxy like nginx to proxy s3g?

**Reply by kerneltime** [reply]
@arghya18 S3G does add a hop in the data path, but if you are trying to channel all the load through a single S3G it will not be sufficient. Like any RESTful service S3G is expected to load balance across multiple instances behind a client-side proxy or a loadbalancer.

**Comment by kerneltime** [comment]
@cloudcheflabs can you elaborate on what the error is? Are you getting `ofs://` not detected as a filesystem?

**Reply by cloudcheflabs** [reply]
@kerneltime 
I cannot reproduce the errors at the moment, but I am trying to tell you what I did.
- I wanted to add iceberg catalog to trino, so first, I copied ozone filesystem jar file to iceberg plugin directory in trino.
- But trino could not know 'ofs://' scheme.

So, for my deployment, I prefer to use ozone as s3 object storage, so I deployed 3 s3g behind nginx proxy, then I added iceberg catalog to trino like this.
```
connector.name=iceberg
hive.metastore.uri=thrift://chango-private-2.chango.private:9083
hive.s3.aws-access-key=ozone/chango-private-3.chango.private@CHANGO.PRIVATE
hive.s3.aws-secret-key=277cce0759a0d6b6ddabf2e5ec7daxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
hive.s3.endpoint=http://chango-private-2.chango.private:80
hive.s3.region=us-east-1
hive.s3.path-style-access=true
hive.s3.ssl.enabled=true
```
`hive.s3.endpoint=http://chango-private-2.chango.private:80` is the address of nginx proxy which will proxy backends 3 s3gs in my ozone cluster.

I am playing with ozone as s3 compatible object storage without problem for now.
